File: src/embeddings.rs
use anyhow::Result;
use rig::client::Nothing;
use rig::client::EmbeddingsClient;
use rig::providers::ollama;
use rig::embeddings::EmbeddingModel;
use rig::embeddings::Embedding;

/// Creates an Ollama client with explicit type annotation for the HTTP client.
fn create_ollama_client() -> Result<ollama::Client<reqwest::Client>> {
    let client: ollama::Client<reqwest::Client> = ollama::Client::new(Nothing)?;
    Ok(client)
}

/// Generates embedding vectors for a batch of text chunks using Ollama's embedding model.
///
/// Returns a vector of embedding vectors (Vec<f32>).
pub async fn embed_texts(texts: &[String]) -> Result<Vec<Vec<f32>>> {
    let client = create_ollama_client()?;
    let model_name =
        std::env::var("EMBEDDING_MODEL").unwrap_or_else(|_| "all-minilm".to_string());

    let emb_model = client.embedding_model(&model_name);

    let embeddings: Vec<Embedding> = emb_model.embed_texts(texts.to_vec()).await?;

    let vectors: Vec<Vec<f32>> = embeddings
        .into_iter()
        .map(|e| e.vec.iter().map(|&v| v as f32).collect())
        .collect();

    Ok(vectors)
}

/// Generates a single embedding vector for a query string.
pub async fn embed_query(query: &str) -> Result<Vec<f32>> {
    let client = create_ollama_client()?;
    let model_name =
        std::env::var("EMBEDDING_MODEL").unwrap_or_else(|_| "all-minilm".to_string());

    let emb_model = client.embedding_model(&model_name);

    let result = emb_model.embed_text(query).await?;

    Ok(result.vec.iter().map(|&v| v as f32).collect())
}

File: python/rusty_rag/embeddings.py
"""Ollama embedding generation."""

import os
import ollama


def embed_texts(texts: list[str], model: str | None = None) -> list[list[float]]:
    """Generate embedding vectors for a batch of text chunks.

    Uses Ollama's embedding API with batch support for efficiency.
    """
    model = model or os.getenv("EMBEDDING_MODEL", "all-minilm")
    response = ollama.embed(model=model, input=texts)
    return response["embeddings"]


def embed_query(query: str, model: str | None = None) -> list[float]:
    """Generate a single embedding vector for a query string."""
    model = model or os.getenv("EMBEDDING_MODEL", "all-minilm")
    response = ollama.embed(model=model, input=query)
    return response["embeddings"][0]
